plugins/catalog-backend-module-bitbucket-server/src/providers/BitbucketServerEntityProvider.ts

```diff
@@ -213,4 +240,332 @@ export class BitbucketServerEntityProvider implements EntityProvider {
+    existing.forEach(item => {
+      if (
+        targets.find(
+          value =>
+            value.metadata.annotations![this.TARGET_ANNOTATION] ===
+            item.spec.target,
+        )
+      ) {
+        stillExisting.push(item);
+      } else {
+        removed.push({
+          locationKey: this.getProviderName(),
+          entity: item,
+        });
+      }
+    });
+
+    const promises: Promise<void>[] = [
+      this.connection.refresh({
+        keys: stillExisting.map(entity => `url:${entity.spec.target}`),
+      }),
+    ];
+
+    if (added.length > 0 || removed.length > 0) {
+      promises.push(
+        this.connection.applyMutation({
+          type: 'delta',
+          added: added,
+          removed: removed,
+        }),
+      );
+    }
+
+    await Promise.all(promises);
+    if (this.eventBroker !== undefined) {
+      await this.publishCatalogMutate({ repoUrl: repoUrl });
+    }
+  }
+
+  /**
+   * Publishes event upon catalog mutation
+   * @param payload Payload to be published to the event broker
+   */
+  private async publishCatalogMutate(
+    payload: Record<string, string>,
+  ): Promise<void> {
+    await new Promise(resolve => {
+      setTimeout(resolve, 10000);
+    });
+    await this.eventBroker!.publish({
+      topic: 'CatalogMutate',
+      eventPayload: payload,
+    });
+  }
+
+  /**
+   * Gets the location entities that are to be newly added to the catalog.
+   * @param targets Location entities for catalog files in the repository that was pushed
+   * @param existing The location entities in the repository that was pushed that already exist
+   * @returns Returns all deferred entities that represent location entities that don't exist in the catalog yet
+   */
+  private getAddedEntities(
+    targets: Entity[],
+    existing: LocationEntity[],
+  ): DeferredEntity[] {
+    const added: DeferredEntity[] = toDeferredEntities(
+      targets.filter(
+        target =>
+          !existing.find(
+            item =>
+              item.spec.target ===
+              target.metadata.annotations![this.TARGET_ANNOTATION],
+          ),
+      ),
+      this.getProviderName(),
+    );
+    return added;
+  }
+
+  /**
+   * Finds all location entities in the catalog that already have the annotation `metadata.annotations.${this.config.host}/repo-url`
+   * that is equivalent to @param repoURL\.
+   * @param repoURL URL for the reposity that the method finds the existing location entities for
+   * @param token Token from class token manager
+   */
+  private async findExistingLocations(
+    catalogRepoUrl: string,
+    token: string,
+  ): Promise<LocationEntity[]> {
+    const filter: Record<string, string> = {};
+    filter.kind = 'Location';
+    filter[`metadata.annotations.${this.TARGET_ANNOTATION}`] = catalogRepoUrl;
+
+    return this.catalogApi!.getEntities({ filter }, { token }).then(
+      result => result.items,
+    ) as Promise<LocationEntity[]>;
+  }
+
+  /**
```

Comment:
seems like this was intended to be added to the `toDeferredEntities` function below, but is a bit misplaced. Or the function is.